### PR TITLE
Add command LSET

### PR DESCRIPTION
**Pull request by tjzhang-BQ**
Adding command LSet

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java [hunk 1]
```diff
@@ -482,4 +482,24 @@ CompletableFuture<Map<String, String[]>> blmpop(
      */
     CompletableFuture<Map<String, String[]>> blmpop(
             String[] keys, PopDirection direction, double timeout);
+
+    /**
+     * Sets the list element at <code>index</code> to <code>element</code>. For details on the index
+     * argument, see {@link #lindex(String, long)}. Throws an exception for out of range <code>index
+     * </code>es.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:489]: See LINDEX for this comment:
```
     * The index is zero-based, so <code>0</code> means the first element, <code>1</code> the second
     * element and so on. Negative indices can be used to designate elements starting at the tail of
     * the list. Here, <code>-1</code> means the last element, <code>-2</code> means the penultimate
     * and so forth.
```

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java [hunk 2]
```diff
@@ -482,4 +482,24 @@ CompletableFuture<Map<String, String[]>> blmpop(
...
+     *
+     * @see <a href="https://valkey.io/commands/lset/">valkey.io</a> for details.
+     * @param key The key of the list.
+     * @param index The index of the element in the list to be set.
+     * @return A simple string reply: OK.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:494]:
```suggestion
     * @return Always <code>OK</code>.<br>
```

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java [hunk 3]
```diff
@@ -482,4 +482,24 @@ CompletableFuture<Map<String, String[]>> blmpop(
...
+     * @example
+     *     <pre>{@code
+     * client.lpush("testKey", new String[] {"three", "one"}).get();
+     * String response = client.lset("testKey", 1, "two").get();
+     * assertEquals(response, "OK");
+     * String[] lrange_result = client.lrange("testKey", 0, -1).get();
+     * assertArrayEquals(lrange_result, new String[] {"one", "two"});
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:501]:
```suggestion
     * assertArrayEquals(lrange_result, new String[] {"three", "two"});
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 4]
```diff
@@ -3723,6 +3724,22 @@ public T lmpop(@NonNull String[] keys, @NonNull PopDirection direction) {
...
 
+    /**
+     * Sets the list element at <code>index</code> to <code>element</code>. For details on the index
+     * argument, see {@link #lindex(String, long)}. Throws an exception for out of range <code>index
+     * </code>es.
+     *
+     * @see <a href="https://valkey.io/commands/lset/">valkey.io</a> for details.
+     * @param key The key of the list.
+     * @param index The index of the element in the list to be set.
+     * @return Command Response: A simple string reply: OK
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3735]:
```suggestion
     * @return Command Response - Always <code>OK</code>.
```

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 5]
```diff
@@ -1543,4 +1544,10 @@ public CompletableFuture<Map<String, String[]>> blmpop(
                 arguments,
                 response -> castMapOfArrays(handleMapOrNullResponse(response), String.class));
     }
+
+    @Override
+    public CompletableFuture<String> lset(String key, long index, String element) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:1549]: nonnull

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 6]
```diff
@@ -5101,4 +5102,30 @@ public void lmpop_with_count_returns_success() {
...
+
+    @SneakyThrows
+    @Test
+    public void lset_returns_success() {
+        // setup
+        String key = "testKey";
+        long index = 0;
+        String element = "two";
+        String[] arguments = new String[] {key, "0", element};
+        String value = "OK";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:5114]:
```suggestion

```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 7]
```diff
@@ -5101,4 +5102,30 @@ public void lmpop_with_count_returns_success() {
...
+        when(commandManager.<String>submitNewCommand(eq(LSet), eq(arguments), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<String> response = service.lset(key, index, element);
+        String payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertEquals(value, payload);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:5129]:
```suggestion
        assertEquals(OK, payload);
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 8]
```diff
@@ -5101,4 +5102,30 @@ public void lmpop_with_count_returns_success() {
...
+    public void lset_returns_success() {
+        // setup
+        String key = "testKey";
+        long index = 0;
+        String element = "two";
+        String[] arguments = new String[] {key, "0", element};
+        String value = "OK";
+
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:5117]:
```suggestion
        testResponse.complete(OK);
```

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java (hunk 1 above)
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:489]: I think it is enough
```suggestion
     * Sets the list element at <code>index</code> to <code>element</code>.
```

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java (hunk 2 above)
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:494]:
```suggestion
     * @return <code>OK</code>.
```

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java (hunk 3 above)
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:501]:
```suggestion
     * String response = client.lset("testKey", 1, "two").get();
     * assertEquals(response, "OK");
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 4 above)
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3735]:
```suggestion
     * @return Command Response - <code>OK</code>.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 9]
```diff
@@ -3723,6 +3724,22 @@ public T lmpop(@NonNull String[] keys, @NonNull PopDirection direction) {
...
+     */
+    public T lset(String key, long index, String element) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3737]: nonnull

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 10]
```diff
@@ -3723,6 +3724,22 @@ public T lmpop(@NonNull String[] keys, @NonNull PopDirection direction) {
         return getThis();
     }
 
+    /**
+     * Sets the list element at <code>index</code> to <code>element</code>. For details on the index
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3728]: same

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 11]
```diff
@@ -4207,4 +4207,38 @@ public void blmpop_timeout_check(BaseClient client) {
...
+        ExecutionException noSuchKeyException =
+                assertThrows(
+                        ExecutionException.class, () -> client.lset(nonExistingKey, index, element).get());
+        assertInstanceOf(RequestException.class, noSuchKeyException.getCause());
+
+        // pushing elements to list
+        client.lpush(key, lpushArgs).get();
+
+        // index out of range
+        ExecutionException indexOutOfBoundException =
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4234]: Add a test with negative index. Does -1 causes an error or replaces the last element?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 12]
```diff
@@ -4207,4 +4207,38 @@ public void blmpop_timeout_check(BaseClient client) {
...
+                assertThrows(ExecutionException.class, () -> client.lset(key, oobIndex, element).get());
+        assertInstanceOf(RequestException.class, indexOutOfBoundException.getCause());
+
+        // assert lset result
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4238]: can you test for a negative index too?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 13]
```diff
@@ -4207,4 +4207,38 @@ public void blmpop_timeout_check(BaseClient client) {
...
+        String response = client.lset(key, index, element).get();
+        assertEquals(OK, response);
+        String[] updatedList = client.lrange(key, 0, -1).get();
+        assertArrayEquals(updatedList, expectedList);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4242]: I think the expected list at this point should be: 
```
        String[] expectedList = {"zero", "three", "two", "one"};
```
since you're only modifying the value at index 0

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 14]
```diff
@@ -250,6 +250,7 @@ private static Object[] listCommands(BaseTransaction<?> transaction) {
         String listKey2 = "{ListKey}-2-" + UUID.randomUUID();
         String listKey3 = "{ListKey}-3-" + UUID.randomUUID();
         String listKey4 = "{ListKey}-4-" + UUID.randomUUID();
+        String listKey5 = "{ListKey}-4-" + UUID.randomUUID();
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:253]: Maybe you can reuse an existing list, if not:
```suggestion
        String listKey5 = "{ListKey}-5-" + UUID.randomUUID();
```

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 15]
```diff
@@ -298,6 +302,9 @@ private static Object[] listCommands(BaseTransaction<?> transaction) {
                     4L, // linsert(listKey3, AFTER, value2, value2)
                     new String[] {listKey3, value3}, // blpop(new String[] { listKey3 }, 0.01)
                     new String[] {listKey3, value1}, // brpop(new String[] { listKey3 }, 0.01)
+                    2L, // lpush(listKey5, new String[] {value2, value3})
+                    OK, // lset(listKey5, 0, value1)
+                    new String[] {value1, value2} // lrange(listKey5, 0, -1)
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:307]:
```suggestion
                    new String[] {value1, value3} // lrange(listKey5, 0, -1)
```
aren't we changing the value at index 0 only?

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 11 above)
Comment (tjzhang-BQ) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:4234]: last element

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 13 above)
Comment (tjzhang-BQ) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:4242]: Lpush is LIFO

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 15 above)
Comment (tjzhang-BQ) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:307]: since lpush is LIFO the list looks like `{value3, value2}` after first being pushed to, and after setting index 0 it becomes `{value1, value2}`

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 16]
```diff
@@ -1543,4 +1544,11 @@ public CompletableFuture<Map<String, String[]>> blmpop(
                 arguments,
                 response -> castMapOfArrays(handleMapOrNullResponse(response), String.class));
     }
+
+    @Override
+    public CompletableFuture<String> lset(
+            @NonNull String key, @NonNull long index, @NonNull String element) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:1550]: nonnull for `long` is an overkill though

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 17]
```diff
@@ -3723,6 +3724,20 @@ public T lmpop(@NonNull String[] keys, @NonNull PopDirection direction) {
         return getThis();
     }
 
+    /**
+     * * Sets the list element at <code>index</code> to <code>element</code>.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3728]:
```suggestion
     * Sets the list element at <code>index</code> to <code>element</code>.
```
and copy doc for `index`

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 18]
```diff
@@ -3723,6 +3724,20 @@ public T lmpop(@NonNull String[] keys, @NonNull PopDirection direction) {
...
+     *
+     * @see <a href="https://valkey.io/commands/lset/">valkey.io</a> for details.
+     * @param key The key of the list.
+     * @param index The index of the element in the list to be set.
+     * @return Command Response - <code>OK</code>.
+     */
+    public T lset(@NonNull String key, @NonNull long index, @NonNull String element) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3735]: same

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/client/src/main/java/glide/api/BaseClient.java (hunk 16 above)
Comment (tjzhang-BQ) [reply java/client/src/main/java/glide/api/BaseClient.java]: true

**Review by Yury-Fridlyand — APPROVED** [review]
